programs/drift/src/instructions/keeper.rs

```diff
@@ -2930,6 +2947,376 @@ pub fn handle_pause_spot_market_deposit_withdraw(
+            msg!(
+                "Skipping settling perp market {} to dlp because oracle slot is not up to date",
+                perp_market.market_index
+            );
+            continue;
+        }
+
+        let fee_pool_token_amount = get_token_amount(
+            perp_market.amm.fee_pool.scaled_balance,
+            &quote_market,
+            perp_market.amm.fee_pool.balance_type(),
+        )?;
+
+        let net_pnl_pool_token_amount = get_token_amount(
+            perp_market.pnl_pool.scaled_balance,
+            &quote_market,
+            perp_market.pnl_pool.balance_type(),
+        )?
+        .cast::<i128>()?
+        .safe_sub(calculate_net_user_pnl(
+            &perp_market.amm,
+            cached_info.oracle_price,
+        )?)?;
+
+        let amm_amount_available =
+            net_pnl_pool_token_amount.safe_add(fee_pool_token_amount.cast::<i128>()?)?;
+
+        if cached_info.last_net_pnl_pool_token_amount == 0
+            && cached_info.last_fee_pool_token_amount == 0
+        {
+            cached_info.last_fee_pool_token_amount = fee_pool_token_amount;
+            cached_info.last_net_pnl_pool_token_amount = net_pnl_pool_token_amount;
+            continue;
+        }
+
+        validate!(
+            perp_market.oracle_id() == cached_info.oracle_id()?,
+            ErrorCode::DefaultError,
+            "oracle id mismatch between amm cache and perp market"
+        )?;
+
+        // Actually transfer the pnl to the lp usdc constituent account
+        validate_market_within_price_band(&perp_market, state, cached_info.oracle_price)?;
+
+        if perp_market.is_operation_paused(PerpOperation::SettlePnl) {
+            msg!(
+                "Cannot settle pnl under current market = {} status",
+                perp_market.market_index
+            );
+
+            return Err(ErrorCode::InvalidMarketStatusToSettlePnl.into());
+        }
+
+        let mint = *ctx.accounts.mint.clone();
+
+        if perp_market.lp_fee_transfer_scalar == 0 {
+            msg!(
+                "lp_fee_transfer_scalar is 0 for perp market {}. Cannot settle pnl",
+                perp_market.market_index
+            );
+            continue;
+        }
+
+        let amount_to_send = amm_amount_available
+            .abs_diff(cached_info.get_last_available_amm_balance()?)
+            .safe_div_ceil(perp_market.lp_fee_transfer_scalar as u128)?
+            .cast::<u64>()?;
+        if amm_amount_available < cached_info.get_last_available_amm_balance()? {
+            let amount_to_send = if amount_to_send > constituent_token_account.amount {
+                cached_info.quote_owed_from_lp +=
+                    amount_to_send.saturating_sub(constituent_token_account.amount) as i64;
+                constituent_token_account.amount
+            } else {
+                amount_to_send
+            };
+
+            controller::token::send_from_program_vault(
```

Comment:
the cpi should be separated from the logic doing the math determining the transfering amount so we can write cargo tests